File: cricket/src/problem/Raintrap.java
package problem;

public class Raintrap {
    
    public static int rainwtuni(int[] arr) {
        int n = arr.length;
        if (n == 0) return 0;
        
        int[] left = new int[n];
        int[] right = new int[n];
        
        // Fill left array
        left[0] = arr[0];
        for (int i = 1; i < n; i++) {
            left[i] = Math.max(left[i - 1], arr[i]);
        }
        
        // Fill right array
        right[n - 1] = arr[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            right[i] = Math.max(right[i + 1], arr[i]);
        }
        
        // Calculate trapped water
        int sum = 0;
        for (int i = 0; i < n; i++) {
            sum += Math.min(left[i], right[i]) - arr[i];
        }
        
        return sum;
    }
    public static void main(String[] args) {
        int[] arr = {3, 0, 2, 0, 4};
        System.out.println(rainwtuni(arr)); // Output: 7
        /*
         *     #
         * #   #
         * # # #
         * # # #*/
    }
    
}

File: Pyhton/Problems/Raintrap.py
def raintrap(arr):
    length = len(arr)
    if length == 0:
        return 0

    left = [0] * length
    right = [0] * length

    # Fill left max array
    left[0] = arr[0]
    for i in range(1, length):
        left[i] = max(left[i - 1], arr[i])

    # Fill right max array
    right[length - 1] = arr[length - 1]
    for i in range(length - 2, -1, -1):
        right[i] = max(right[i + 1], arr[i])

    # Calculate trapped water
    total_water = 0
    for i in range(length):
        total_water += min(left[i], right[i]) - arr[i]

    return total_water
height = [0,1,0,2,1,0,1,3,2,1,2,1]
print(raintrap(height))  # Output: 6
